Fix #694: Lowfi: HelpActivity (Part 2) - FAQ [Blocked part-1]

SayantanBanerjee16 opened the pull request:
<!--
  - Thanks for submitting code to Oppia! Please fill out the following as part of
  - your pull request so we can review your code more easily.
  -->

## Explanation
<!--
  - Explain what your PR does. If this PR fixes an existing bug, please include
  - "Fixes #bugnum:" in the explanation so that GitHub can auto-close the issue
  - when this PR is merged.
  -->
Fixes #694: 

Created a recycler view inside the FAQ activity using MVVM architecture and populated it with the items fetching from a string-array declared in array.xml resource. Related tests also have been implemented.

## Checklist
<!-- Please tick the relevant boxes by putting an "x" in them. -->
- [x] The PR title starts with "Fix #bugnum: ", followed by a short, clear summary of the changes. (If this PR fixes part of an issue, prefix the title with "Fix part of #bugnum: ...".)
- [x] The PR explanation includes the words "Fixes #bugnum: ..." (or "Fixes part of #bugnum" if the PR only partially fixes an issue).
- [x] The PR follows the [style guide](https://github.com/oppia/oppia-android/wiki/Coding-style-guide).
- [x] The PR does not contain any unnecessary auto-generated code from Android Studio.
- [x] The PR is made from a branch that's **not** called "develop".
- [x] The PR is made from a branch that is up-to-date with "develop".
- [x] The PR's branch is based on "develop" and not on any other branch.
- [x] The PR is **assigned** to an appropriate reviewer in both the **Assignees** and the **Reviewers** sections.

SayantanBanerjee16 committed:
FAQ recycler view set up

SayantanBanerjee16 committed:
FAQ questions set up

SayantanBanerjee16 committed:
Search button enabled in toolbar

SayantanBanerjee16 committed:
Back button enabled in toolbar

SayantanBanerjee16 committed:
FAQ Recycler view layout formatting done

SayantanBanerjee16 committed:
FAQ Fragment tests added

SayantanBanerjee16 committed:
NIT changes

SayantanBanerjee16 commented:
@rt4914 Please review this PR. Thanks :)

SayantanBanerjee16 committed:
NIT changes

rt4914 reviewed (changes requested):
PR looks good at first look but there is implementation change.

If you see the mocks in detail, the `header` also scrolls with the items, so that means that `Featrued Questions` is `0th` index of recyclerview. Therefore you will need to inflate multiple types of views in RecyclerView, it has been done at various places in the app.

SayantanBanerjee16 committed:
Item Header xml created

SayantanBanerjee16 committed:
Recycler view attachment with multiple views done and related test written

SayantanBanerjee16 committed:
Merge develop

SayantanBanerjee16 committed:
NIT fix

SayantanBanerjee16 commented:
> PR looks good at first look but there is implementation change.
> 
> If you see the mocks in detail, the `header` also scrolls with the items, so that means that `Featrued Questions` is `0th` index of recyclerview. Therefore you will need to inflate multiple types of views in RecyclerView, it has been done at various places in the app.

@rt4914 I have implemented multiple types of Views inside RecyclerView, one for header and another for content. Its working fine and test also implemented. PTAL. Thanks :)

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQActivity.kt:
```diff
@@ -16,7 +17,12 @@ class FAQActivity : InjectableAppCompatActivity() {
     activityComponent.inject(this)
     faqActivityPresenter.handleOnCreate()
-    title = getString(R.string.frequently_asked_questions)
+    title = getString(R.string.FAQs)
```
Do not set title here. I know you have used `<include>` in xml and because of which you are using this code. But instead I would suggest you to checkout the of `exploration_activity` and create toolbar in that way. Also, we should include data binding in this file at this stage only.

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQActivity.kt:
```diff
@@ -25,4 +31,9 @@ class FAQActivity : InjectableAppCompatActivity() {
     }
   }
+
+  override fun onSupportNavigateUp(): Boolean {
```
Instead of this, checkout the implementation of back button in `StoryActivity`.

SayantanBanerjee16 replied:
@rt4914 Can you tell me a specific position where should I look for? I searched but didn't found anything related to the back button.

SayantanBanerjee16 replied:
@rt4914 Can you give me a proper location where to look at as I Can't find any reference of Back Button in Story Activity.

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQActivityPresenter.kt:
```diff
@@ -10,9 +12,17 @@ import javax.inject.Inject
+    val toolbar = activity.findViewById<View>(R.id.faq_activity_toolbar) as Toolbar
+    activity.setSupportActionBar(toolbar)
+    activity.supportActionBar!!.setDisplayShowHomeEnabled(true)
+    activity.supportActionBar!!.setDisplayHomeAsUpEnabled(true);
```
Remove `;` from the end.

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQActivityPresenter.kt:
```diff
@@ -10,9 +12,17 @@ import javax.inject.Inject
   fun handleOnCreate() {
     activity.setContentView(R.layout.faq_activity)
+    setUpToolbar()
     activity.supportFragmentManager.beginTransaction().add(
```
Always check if the fragment is already present or not before adding it.

Check the implementation in `MyDownloadsActivityPresenter` where we first check if fragment is present or not. Also, we need to fix this same thing in `HelpActivityPresenter` also.

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQCategoryAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+  private val arrayList: ArrayList<FAQViewModel>
+) : RecyclerView.Adapter<RecyclerView.ViewHolder>() {
+
+  private val header = 1
```
Create enum instead of this. Check enum in `StoryFragmentPresenter`.

SayantanBanerjee16 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQCategoryAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+
+  override fun getItemViewType(position: Int): Int {
+    return if (position == 0) {
+      header;
```
Remove `;` here and everywhere else.

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQCategoryAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+      val view: View = LayoutInflater.from(parent.context).inflate(
+        R.layout.faq_item_header,
+        parent,
+        /* attachToParent= */false
```
Space issue:
`/* attachToParent= */ false`

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQCategoryAdapter.kt:
```diff
@@ -0,0 +1,68 @@
+    }
+  }
+}
+
```
Remove this extra blank line

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQFragmentPresenter.kt:
```diff
@@ -3,21 +3,42 @@ package org.oppia.app.help.faq
-    val binding = FaqFragmentBinding.inflate(inflater, container, /* attachToRoot= */ false)
-    // NB: Both the view model and lifecycle owner must be set in order to correctly bind LiveData elements to
-    // data-bound view models.
+    val binding = FaqFragmentBinding.inflate(
```
Single line:
`val binding = FaqFragmentBinding.inflate(inflater, container,/* attachToRoot= */ false)`

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQViewModel.kt:
```diff
@@ -0,0 +1,6 @@
+import org.oppia.app.viewmodel.ObservableViewModel
+
+/** [ObservableViewModel] for the recycler view of FAQActivity. */
+class FAQViewModel(val title: String) : ObservableViewModel()
```
Instead of `ObservableViewModel ` make it `ViewModel ` only because the string won't change.

rt4914 replied:
Update comment also.

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/res/layout/faq_activity.xml:
```diff
@@ -1,7 +1,24 @@
+    android:layout_height="match_parent"
+    android:orientation="vertical">
+
+    <include
```
The implementation in this should be similar to `exploration_activity`

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/res/layout/faq_fragment.xml:
```diff
@@ -6,24 +6,18 @@
-        app:titleTextColor="@color/white" />
-    </com.google.android.material.appbar.AppBarLayout>
+        android:layout_height="match_parent" />
+    </LinearLayout>
```
Press `Control + Alt + L` - reformatting shortcut to fix the formatting.

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/res/layout/faq_item.xml:
```diff
@@ -0,0 +1,42 @@
+<layout xmlns:android="http://schemas.android.com/apk/res/android"
+  xmlns:app="http://schemas.android.com/apk/res-auto">
+
+  <data class="FAQItemBinding">
```
This should be
`<data>`

SayantanBanerjee16 replied:
Done

rt4914 commented on app/src/main/res/layout/faq_item.xml:
```diff
@@ -0,0 +1,42 @@
+  <data class="FAQItemBinding">
+
+    <variable
+      name="viewmodel"
```
Rename:
`viewmodel ` to `viewModel`

SayantanBanerjee16 replied:
Done

SayantanBanerjee16 replied:
done

rt4914 commented on app/src/main/res/layout/faq_item.xml:
```diff
@@ -0,0 +1,42 @@
+        android:textSize="16sp"
+        android:textStyle="bold" />
+
+      <View
```
This line should not be visible for the last index.

SayantanBanerjee16 replied:
done.

rt4914 commented on app/src/main/res/values/dimens.xml:
```diff
@@ -15,6 +15,7 @@
   <dimen name="elevation_0dp">0dp</dimen>
   <dimen name="padding_72dp">72dp</dimen>
   <dimen name="padding_24dp">24dp</dimen>
+  <dimen name="margin_24dp">24dp</dimen>
```
Instead of introducing these padding here, prefer directly hard-coding them where ever you are using them

SayantanBanerjee16 replied:
Done

rt4914 reviewed (changes requested):
Thanks. Suggested changes.

SayantanBanerjee16 committed:
NIT chnages

SayantanBanerjee16 committed:
Enum created in CategoryAdapter

SayantanBanerjee16 committed:
Fragments checked null or not null. Created only when null

SayantanBanerjee16 committed:
FAQ Activity attached to a View Model

SayantanBanerjee16 committed:
FAQ Activity back button and title set inside presenter

SayantanBanerjee16 committed:
Divider listener added

SayantanBanerjee16 committed:
NIT fix

SayantanBanerjee16 committed:
Bug fix in items of recycler view

SayantanBanerjee16 committed:
Binding change from FAQItemBinding to FAQContentBinding

SayantanBanerjee16 committed:
Tests implemented

SayantanBanerjee16 commented:
@rt4914 Implemented Enum in Adapter, Ensured Divider Line doesn't appear in the last item (Related Test had been written) and also introduced Data Binding to FAQ Activity. PTAL. Thanks :)

1) The test which I have written hard codes the number of items on the list. Is there a way in which I can get dynamically the item count of the last. If so, how?
2) The back button is enabled but not reflecting.

Rest everything working fine.

rt4914 commented on app/src/main/java/org/oppia/app/help/HelpActivityPresenter.kt:
```diff
@@ -17,10 +17,12 @@ class HelpActivityPresenter @Inject constructor(private val activity: AppCompatA
-      R.id.help_fragment_placeholder,
-      HelpFragment()
-    ).commitNow()
+    if (getHelpFragment() == null) {
```
Changes in this file:
Prefer not using null, I missed this in last PR review.
`private lateinit var navigationDrawerFragment: NavigationDrawerFragment`

rt4914 replied:
This will bring change in `setUpNavigationDrawer` also
`navigationDrawerFragment.setUpDrawer(`

SayantanBanerjee16 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQActivityPresenter.kt:
```diff
@@ -1,18 +1,52 @@
+    return activity.supportFragmentManager.findFragmentById(R.id.faq_fragment_placeholder) as FAQFragment?
+  }
+
+  private fun getFAQActivityViewModel(): FAQActivityViewModel {
```
You can remove `FAQActivityViewModel` because it is actually not doing anything, once you remove that file, you can remove its implementation from this file and also the corresponding xml file.

SayantanBanerjee16 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQActivityViewModel.kt:
```diff
@@ -0,0 +1,7 @@
+import javax.inject.Inject
+
+/** [ViewModel] for the recycler view of FAQActivity. */
+class FAQActivityViewModel @Inject constructor() : ViewModel()
```
Remove this file.

SayantanBanerjee16 replied:
Done

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQCategoryAdapter.kt:
```diff
@@ -0,0 +1,78 @@
+import org.oppia.app.databinding.FaqContentBinding
+
+/** The Recycler View adapter in the [FAQFragment]. */
+class FAQCategoryAdapter(
```
This adapter implementation is incorrect. I think now we cannot skip the implementation of bindable adapter anymore considering that even the implementation of adapter is incorrect here. 

The best reference for your FAQ list is `StoryActivity` recyclerview implementation.

You can try this for some time but once you feel that you are not able to figure this out, feel free to inform me, so that I can work on this PR which you can check and learn from.

SayantanBanerjee16 replied:
> This adapter implementation is incorrect. I think now we cannot skip the implementation of bindable adapter anymore considering that even the implementation of adapter is incorrect here.
> 
> The best reference for your FAQ list is `StoryActivity` recyclerview implementation.
> 
> You can try this for some time but once you feel that you are not able to figure this out, feel free to inform me, so that I can work on this PR which you can check and learn from.

Okay I will start working on this.

SayantanBanerjee16 replied:
done :)

rt4914 reviewed (changes requested):
The adapter implementation is still incorrect. So I suggest trying your hands around the bindable adapter similar to `StoryActivity`.

SayantanBanerjee16 committed:
Unused code removed

SayantanBanerjee16 committed:
Bindable Adapter declaration done

SayantanBanerjee16 committed:
Nit fix

SayantanBanerjee16 commented:
@rt4914 So, I deleted the Adapter and started the implementation of setting up BindableAdapter in FAQFragmentPresenter, in which I have created a superclass `FAQItemViewModel` which is extended by `FAQHeaderViewModel` (Header View Model) and `FAQContentViewModel` (Contents View Model).

Now the recycler view is set up. PTAL at `FAQFragmentPresenter.kt` . Now please help me out in setting the data and also setting `position = 0` goes to the `FAQHeaderViewModel` and rest positions go to the `FAQContentViewModel`. Thanks :)

rt4914 reviewed (changes requested):
Please take latest merge from develop as your branch is running behind that code.

Take latest merge from develop branch.

One main point is that you have implemented the classes very close to correct but while take code from other files you have't focused on xml files. Please checkout the xml files also while take reference from some other activity/fragment.

I can just give you hints but you will have to figure out where everything goes.

```
val faqItemList: List<FAQItemViewModel> by lazy {
    getRecyclerViewItemList()
  }
```

Set your faqItemList in your recyclerview using `databinding`. For that first you will need to bind your presenter to your `xml` and in your xml you will need to use `faqItemList` to set data to your recyclerview. For more details check `story_fragment`. 
[NOTE: This will only work once you merge your branch with latest develop because as per your implementation I have made some changes to develop.]

Also, while copying code you are copying irrelevant code also, please also think before copying anything.

SayantanBanerjee16 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into faq-content

rt4914 commented:
@SayantanBanerjee16 Haven't seen any updates in this, if you are stuck please feel free to inform.

SayantanBanerjee16 commented:
> @SayantanBanerjee16 Haven't seen any updates in this, if you are stuck please feel free to inform.

Yes sir, I am having a hard time finding the right way to introduce data to Bindable Adapter. I want to invest today's day also if I can figure it out with something.

Meanwhile, can you assign me some other issue too so that I can work on both of these parallelly?

rt4914 commented:
> > @SayantanBanerjee16 Haven't seen any updates in this, if you are stuck please feel free to inform.
> 
> Yes sir, I am having a hard time finding the right way to introduce data to Bindable Adapter. I want to invest today's day also if I can figure it out with something.
> 
> Meanwhile, can you assign me some other issue too so that I can work on both of these parallelly?

Actually I can understand that it can be hard at first. So I think you should focus on this only and not on any other issue. Lets give it one more day, if you are still stuck, do inform me, I will help you out.

Also, please take help from this comment https://github.com/oppia/oppia-android/pull/759#pullrequestreview-373474911

SayantanBanerjee16 committed:
refractor class names

SayantanBanerjee16 committed:
VIew Model introduced, data fetched and displayed done

SayantanBanerjee16 committed:
NIT changes

SayantanBanerjee16 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into faq-content

SayantanBanerjee16 committed:
Bug fix from app:data to app:list

SayantanBanerjee16 committed:
Divider set to false on final value

SayantanBanerjee16 committed:
Back button bug fixed in FAQActivity

SayantanBanerjee16 commented:
> > > @SayantanBanerjee16 Haven't seen any updates in this, if you are stuck please feel free to inform.
> > 
> > 
> > Yes sir, I am having a hard time finding the right way to introduce data to Bindable Adapter. I want to invest today's day also if I can figure it out with something.
> > Meanwhile, can you assign me some other issue too so that I can work on both of these parallelly?
> 
> Actually I can understand that it can be hard at first. So I think you should focus on this only and not on any other issue. Lets give it one more day, if you are still stuck, do inform me, I will help you out.
> 
> Also, please take help from this comment [#759 (review)](https://github.com/oppia/oppia-android/pull/759#pullrequestreview-373474911)

@rt4914 Finally able to do it :) I previously got confused in setting up classes properly but now its resolved. Data set in Recycler View, and also ensured the divider line doesn't occur after the last Item. Tests also implemented. PTAL. Thanks :)

SayantanBanerjee16 commented:
BTW I have no clue why non-flaky-task failed. It should have passed. The error message displaying by hovering on the link `Details` in non_flaky_task is nothing to do with me.

`testSplashActivity_secondOpen_routesToChooseProfileActivity - org.oppia.app.splash.SplashActivityTest`

rt4914 commented:
> BTW I have no clue why non-flaky-task failed. It should have passed. The error message displaying by hovering on the link `Details` in non_flaky_task is nothing to do with me.
> 
> `testSplashActivity_secondOpen_routesToChooseProfileActivity - org.oppia.app.splash.SplashActivityTest`

@SayantanBanerjee16 Don't worry about that. Let's just finalise your implementation first.

rt4914 commented:
> > > > @SayantanBanerjee16 Haven't seen any updates in this, if you are stuck please feel free to inform.
> > > 
> > > 
> > > Yes sir, I am having a hard time finding the right way to introduce data to Bindable Adapter. I want to invest today's day also if I can figure it out with something.
> > > Meanwhile, can you assign me some other issue too so that I can work on both of these parallelly?
> > 
> > 
> > Actually I can understand that it can be hard at first. So I think you should focus on this only and not on any other issue. Lets give it one more day, if you are still stuck, do inform me, I will help you out.
> > Also, please take help from this comment [#759 (review)](https://github.com/oppia/oppia-android/pull/759#pullrequestreview-373474911)
> 
> @rt4914 Finally able to do it :) I previously got confused in setting up classes properly but now its resolved. Data set in Recycler View, and also ensured the divider line doesn't occur after the last Item. Tests also implemented. PTAL. Thanks :)

I know it becomes hard to implement at first but hopefully now you might find it simpler as compared to the RecyclerView Adapter. Thanks :)

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQFragmentPresenter.kt:
```diff
@@ -1,23 +1,113 @@
 ) {
+  private lateinit var binding: FaqFragmentBinding
+  private lateinit var linearLayoutManager: LinearLayoutManager
+  private lateinit var linearSmoothScroller: RecyclerView.SmoothScroller
```
You don't need this `linearSmoothScroller` to remove entire code related to this.

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQFragmentPresenter.kt:
```diff
@@ -1,23 +1,113 @@
+    binding = FaqFragmentBinding.inflate(inflater, container, /* attachToRoot = */ false)
+
+    binding.faqFragmentRecyclerView.apply {
+      layoutManager = linearLayoutManager
```
Remove `linearLayoutManager` and use `LinearLayoutManager(activity.applicationContext)` directly here.

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/FAQFragmentPresenter.kt:
```diff
@@ -1,23 +1,113 @@
+    }
+  }
+
+  private fun dipToPixels(dipValue: Int): Int {
```
Do won't need this function once you remove `linearSmoothScroller `

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/main/java/org/oppia/app/help/faq/faqItemViewModel/FAQContentViewModel.kt:
```diff
@@ -0,0 +1,10 @@
+class FAQContentViewModel(val title: String) : FAQItemViewModel() {
+
+  /** Used to control visibility of divider. */
+  var showDivider = ObservableField(true)
```
Update:

`val showDivider = ObservableField(true)`

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/main/res/layout/faq_activity.xml:
```diff
@@ -1,7 +1,36 @@
+      android:id="@+id/faq_fragment_placeholder"
+      android:layout_width="match_parent"
+      android:layout_height="match_parent"
+      tools:context=".help.faq.FAQActivity" />
```
Remove this line. This is not needed here.

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/main/res/layout/faq_item_header.xml:
```diff
@@ -0,0 +1,37 @@
+      app:layout_constraintTop_toTopOf="parent">
+
+      <TextView
+        android:id="@+id/recycler_item_text_view"
```
Lets name `recycler_item_text_view ` in a better way. Name it as `faq_question_text_view`

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/sharedTest/java/org/oppia/app/faq/FAQFragmentTest.kt:
```diff
@@ -0,0 +1,67 @@
+@RunWith(AndroidJUnit4::class)
+class FAQFragmentTest {
+
+  var itemCount: Int = 9
```
Avoid using `var` as much as possible and instead use `val`

`val itemCount: Int = 9`

SayantanBanerjee16 replied:
Done :)

rt4914 commented on app/src/sharedTest/java/org/oppia/app/faq/FAQFragmentTest.kt:
```diff
@@ -0,0 +1,67 @@
+  fun openFAQActivity_scrollRecyclerViewToZeroPosition_checkFeaturedQuestionsDisplayedSuccessfully() {
+    launch(FAQActivity::class.java).use {
+      onView(withId(R.id.faq_fragment_recycler_view)).perform(
+        RecyclerViewActions.scrollToPosition<RecyclerView.ViewHolder>(
```
Update this to `scrollToPosition<RecyclerView.ViewHolder>(` here and everwhere else and remove unused imports.

Also, please correct the indentation for entire file. You can do this by first bringing the relative code to a single line rather than multiple lines and then press `Control + Alt + L`

SayantanBanerjee16 replied:
Done :)

rt4914 reviewed (changes requested):
Nice implementation. I have suggested some changes. Thanks.

SayantanBanerjee16 committed:
Refactoring done and NIT update

SayantanBanerjee16 committed:
Indentation fixed

SayantanBanerjee16 committed:
Merge branch 'develop' of https://github.com/oppia/oppia-android into faq-content

SayantanBanerjee16 committed:
NIT update

SayantanBanerjee16 commented:
@rt4914 Done with all the refactoring. PTAL Thanks :)

rt4914 reviewed (approved):
LGTM, thanks.
